Potential Crash fix by not propagating empty value for face attribute

Ahmad-S792 opened the pull request:
#### 7f50b6d09b385b564a7e9733687178ac0d2692c8
<pre>
Potential Crash fix by not propagating empty value for face attribute

Potential Crash fix by not propagating empty value for face attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=248434">https://bugs.webkit.org/show_bug.cgi?id=248434</a>

Reviewed by Tim Nguyen.

Merge - <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=190788

This patch is to add check to ensure that &quot;faceAttr&quot; is not null / empty value and such values are not propagated to lead to stability issues (i.e., crashes).

* Source/WebCore/html/HTMLFontElement.cpp:
(HTMLFontElement::collectPresentationalHintsForAttribute): Add check for empty / null values
* LayoutTests/fast/css/font-face-attribute-remove.html: Add Test Case
* LayoutTests/fast/css/font-face-attribute-remove-expected.html: Add Test Case Expectation

Canonical link: <a href="https://commits.webkit.org/257248@main">https://commits.webkit.org/257248@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee4dc5ee17ccb22f9a28cd26bd69b905ad1b10b4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98308 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7516 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31446 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107758 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168025 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102250 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8021 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/36276 "Built successfully") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90882 "Updated gtk dependencies (failure)") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/104370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103961 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84899 "Built successfully") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/36/builds/90882 "Updated gtk dependencies (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87913 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/89625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/36/builds/90882 "Updated gtk dependencies (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1473 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1417 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/22548 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4991 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6311 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2768 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41961 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fast/css/font-face-attribute-remove.html:
```diff
@@ -0,0 +1,5 @@
+<font id="f" face="helvetica"></font>
```
Typically in crash tests and their reference, there is a line saying "PASS if this doesn't crash." or something similar. There is also no `.png` expectation as well

Ahmad-S792 replied:
I just added platform specific 'png' file for GTK where required, do you want me to add it for other platforms as well or tweak test?

nt1m replied:
Can you tweak the test and remove the GTK expectation?

Ahmad-S792 replied:
Done and pushed and also removed GTK Expectation. :-)

nt1m commented on LayoutTests/fast/css/font-face-attribute-remove.html:
```diff
@@ -0,0 +1,18 @@
+    if (window.testRunner)
+    testRunner.notifyDone();
+}, 0);
+</script>
```
```
<p>PASS if it does not crash.</p>
<font id="f" face="helvetica"></font>
<script>
 // attributes[1] is face
 f.attributes[1].textContent = null;
</script>
 ```

That would have been enough, and for the reference file:

```
<p>PASS if it does not crash.</p>
```

Ahmad-S792 replied:
Updated and pushed it.

nt1m commented on LayoutTests/fast/css/font-face-attribute-remove.html:
```diff
@@ -0,0 +1,6 @@
+<p>PASS if it does not crash.</p>
+<font id="f" face="helvetica"></font>
+ <script>
```
there's some extra whitespace before the `<`

Ahmad-S792 replied:
Fixed.

Ahmad-S792 committed:
Potential Crash fix by not propagating empty value for face attribute

Potential Crash fix by not propagating empty value for face attribute
https://bugs.webkit.org/show_bug.cgi?id=248434

Reviewed by Tim Nguyen.

Merge - https://src.chromium.org/viewvc/blink?view=revision&revision=190788

This patch is to add check to ensure that "faceAttr" is not null / empty value and such values are not propagated to lead to stability issues (i.e., crashes).

* Source/WebCore/html/HTMLFontElement.cpp:
(HTMLFontElement::collectPresentationalHintsForAttribute): Add check for empty / null values
* LayoutTests/fast/css/font-face-attribute-remove.html: Add Test Case
* LayoutTests/fast/css/font-face-attribute-remove-expected.html: Add Test Case Expectation

Canonical link: https://commits.webkit.org/257248@main
